File: lib/ts/src/internal/types/TObject.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { TTypeDeclaration } from './TTypeDeclaration.js';
import { ValidationFailure } from '../validation/ValidationFailure.js';
import { TType } from './TType.js';
import { validateObject } from '../validation/ValidateObject.js';
import { generateRandomObject } from '../generation/GenerateRandomObject.js';
import { GenerateContext } from '../generation/GenerateContext.js';
import { ValidateContext } from '../validation/ValidateContext.js';

export const objectName: string = 'Object';

export class TObject implements TType {
    getTypeParameterCount(): number {
        return 1;
    }

    validate(value: any, typeParameters: TTypeDeclaration[], ctx: ValidateContext): ValidationFailure[] {
        return validateObject(value, typeParameters, ctx);
    }

    generateRandomValue(
        blueprintValue: any,
        useBlueprintValue: boolean,
        typeParameters: TTypeDeclaration[],
        ctx: GenerateContext,
    ): any {
        return generateRandomObject(blueprintValue, useBlueprintValue, typeParameters, ctx);
    }

    getName(): string {
        return objectName;
    }
}

File: lib/go/internal/types/TObject.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package types

const objectName = "Object"

// TObject represents the Telepact Object type.
type TObject struct{}

// NewTObject constructs a TObject instance.
func NewTObject() *TObject {
	return &TObject{}
}

// GetTypeParameterCount implements TType.
func (t *TObject) GetTypeParameterCount() int {
	return 1
}

// Validate implements TType.
func (t *TObject) Validate(value any, typeParameters []*TTypeDeclaration, ctx *ValidateContext) []*ValidationFailure {
	return ValidateObject(value, typeParameters, ctx)
}

// GenerateRandomValue implements TType.
func (t *TObject) GenerateRandomValue(blueprintValue any, useBlueprintValue bool, typeParameters []*TTypeDeclaration, ctx *GenerateContext) any {
	return GenerateRandomObject(blueprintValue, useBlueprintValue, typeParameters, ctx)
}

// GetName implements TType.
func (t *TObject) GetName(ctx *ValidateContext) string {
	return objectName
}
